File: python/409_longest_palindrome.py
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        char_count = dict()
        for c in s:
            if char_count.get(c):
                char_count[c] += 1
            else:
                char_count[c] = 1
        
        longest = 0
        singleton = False
        for v in char_count.values():
            if v % 2 == 0:
                longest += v
            else:
                singleton = True
                longest += v - 1
        
        return longest + singleton

File: c++/409_longest_palindrome.cpp
class Solution {
public:
    int longestPalindrome(string s) {
        unordered_map<char, int> char_count;
        int longest = 0;
        bool single = false;

        for (char c : s) {
            if (char_count.find(c) == char_count.end()) {
                char_count[c] = 1;
            } else {
                char_count[c] += 1;
            }
        }

        for (auto& [c, count] : char_count) {
            if (count % 2 == 0) {
                longest += count;
            } else {
                if (single) {
                    longest += count - 1;
                } else {
                    longest += count;
                    single = true;
                }
            }
        }    

        return longest;
    }
};
